Encryption.md

```diff
@@ -132,49 +203,164 @@ the encryption buffer is comprised of an IV, ciphertext and tag, as described in
+  8: optional ColumnCryptoMetaData crypto_meta_data
+}
+```
+
+The `path_in_schema` field in  `EncryptionWithColumnKey` structure is needed because the 
+column name (path) is kept inside the encrypted `ColumnMetaData`. The column name is required 
+to build the AAD for decryption of the `ColumnMetaData` (see above). Also, the column name can 
+be used for explicit passing of a decryption key for this column, without calling a key retrieval callback.
+
+### Encryption key metadata
+A wide variety of services and tools for management of encryption keys exist in the industry 
+today. Public clouds offer different key management services (KMS), and organizational IT 
+systems either build proprietary key managers in-house or adopt open source tools for 
+on-premises deployment. Besides the diversity of management tools, there are many ways 
+to generate and handle the keys themselves (generate Data keys inside KMS – or locally upon 
+data encryption; use Data keys only, or use Master keys to encrypt/wrap the Data keys; store 
+the wrapped key material inside the data file, or at a separate location; etc). There is also 
+a large variety of authorization and certification methods, required to control the access to 
+encryption keys.
+
+Parquet is not limited to a single KMS, key generation/wrapping method, or authorization 
+service. Instead, Parquet provides a developer with a simple interface (key_metadata byte 
+array, and a key retrieval callback) that can be utilized for implementation of any key 
+management scheme. For example, the key_metadata can keep a serialized 
+
+   * String ID of a Data encryption key (for key retrieval from KMS).
+   * Wrapped Data key, and string ID of a Master key.  The Data key is generated randomly and 
+   wrapped either remotely in a KMS, or locally after retrieving the Master key from KMS. 
+   Wrapping format depends on the KMS and can be a JSON string, or a base64 encoded byte array 
+   in a case of local wrapping.
+   * Short ID (counter) of a key inside the Parquet data file. The Data key is wrapped with a Master 
+   key using one of the options described above – but the resulting key material is stored separately, 
+   outside the data file, and will be retrieved using the counter and file path.
+   * Any of the three above, plus a string ID of the KMS instance (in case there are many).
+   * Random string (useful for creation of footer keys in organizations not willing to 
+   authenticate and connect all readers to a KMS. The footer key is generated as F(key_metadata), 
+   where F is an algorithm unknown outside organization. These keys are less secure than 
+   KMS-managed keys, but better than using no keys at all in plaintext footers, see a discussion below).
+
+Key metadata can also be empty (not set) - in a case the encryption keys are fully managed by the 
+caller code, and passed explicitly to Parquet for the footer and each encrypted column.
+
+### Protection of column metadata
```

Comment:
There are two more options I'd like to explore. The problem with the current approach is that it requires more backward seeks to read the data and to make that more efficient, some assumptions about how the encrypted column meta data buffers are laid out (backward seek to the first one, then read to footer start, then alter offsets to read from that buffer).

## Embedding encrypted `ColumnMetaData`

One option is to add a binary field to `ColumnChunk` called `encrypted_column_metadata` and put the encryption buffer there. This would mean we have to decode a Thrift structure after decrypting the data, but it would avoid the location and seek problems.

## Using unencrypted `ColumnMetaData`

The new column index structures encode all of the data from statistics, which is optional and the main piece of data that encryption is protecting. This spec could simply state that column stats must be null when encrypting a column. Without statistics, the most sensitive data is `num_values`, `encodings` / `encoding_stats`, and the `uncompressed_size` (compressed size is the same so you can tell this from the ciphertext).

I think it would be reasonable to expose the remaining data fields because they contain no values. We could also add `num_values` to `OffsetIndex` and create a filtered version of `ColumnMetaData` without statistics or the count.

The impact of this on performance would be low when the column index and offset index structures are written because Parquet could get the column stats from those structures.

Reply:
I like the **Embedding encrypted ColumnMetaData** option - it saves a seek, but still allows for column chunk filtering without a need to compute column (chunk) stats from multiple index structures.

Reply:
Sounds good to me.